src/webgpu/api/operation/adapter/requestDevice.spec.ts

```diff
@@ -1,16 +1,301 @@
+        `Expected ${limit} == default: ${device.limits[limit]} != ${kLimitInfo[limit].default}`
+      );
+    }
+
+    device.destroy();
+  });
+
+g.test('undefined')
+  .desc(
+    `
+    Test that specifying undefined for the device descriptor should result in a default device.
+    - No features listed in device
+    - Default limits`
+  )
+  .fn(async t => {
+    const gpu = getGPU();
+    const adapter = await gpu.requestAdapter();
+    assert(adapter !== null);
+    const device = await adapter.requestDevice(undefined);
+    assert(device !== null);
+
+    // Device should have no features.
+    t.expect(device.features.size === 0, 'Default device should not have any features');
+    // All limits should be defaults 
+    for (var limit of kLimits) {
+      t.expect(
+        device.limits[limit] === kLimitInfo[limit].default,
+        `Expected ${limit} == default: ${device.limits[limit]} != ${kLimitInfo[limit].default}`
+      );
+    }
+
+    device.destroy();
+  });
+
+g.test('invalid')
+  .desc(
+    `
+    Test that requesting device on an invalid adapter resolves with lost device.
+    - Induce invalid adapter via a device lost from a device.destroy()`
+  )
+  .fn(async t => {
+    const gpu = getGPU();
+    const adapter = await gpu.requestAdapter();
+    assert(adapter !== null);
+
+    {
+      // Request a device and destroy it immediately afterwards.
+      const device = await adapter.requestDevice();
+      assert(device !== null);
+      device.destroy();
+      const lostInfo = await device.lost;
+      t.expect(lostInfo.reason === 'destroyed');
+    }
+
+    // The adapter should now be invalid since a device was lost. Requesting another device should
+    // return an already lost device.
+    var kTimeoutMS = 1000;
+    const device = await adapter.requestDevice();
+    assert(device !== null);
+    t.shouldResolve(raceWithRejectOnTimeout(device.lost, kTimeoutMS, 'device was not lost'));
+  });
+
+g.test('features,unknown')
+  .desc(
+    `
+    Test requesting device with an unknown feature.`
+  )
+  .fn(async t => {
+    const gpu = getGPU();
+    const adapter = await gpu.requestAdapter();
+    assert(adapter !== null);
+
+    let feature: any = "unknown-feature";
+    t.shouldReject(
+      'TypeError',
+      adapter.requestDevice({ requiredFeatures: [feature] })
+    );
+  });
+
+g.test('features,known')
+  .desc(
+    `
+    Test requesting device with all features.
+    - Succeeds with device supporting feature if adapter supports the feature.
+    - Rejects if the adapter does not support the feature.`
+  )
+  .params(u =>
+    u.combine('feature', kFeatureNames)
+  )
+  .fn(async t => {
+    const { feature } = t.params;
+
+    const gpu = getGPU();
+    const adapter = await gpu.requestAdapter();
+    assert(adapter !== null);
+
+    if (adapter.features.has(feature)) {
+      const device = await adapter.requestDevice({ requiredFeatures: [feature] });
+      assert(device !== null);
```

Comment:
nit: Assert shouldn't be necessary because device isn't supposed to be allowed to be null.
(If it is anyway, the test will crash on the next line on `device.features`.)